### ISSUE TITLE
Sketch: Limit detrimental effect of softmax temp exploration on value

### ISSUE THREAD

**Issue by Naphthalin** [LeelaChessZero/lc0#1308]
_This is a documentation of a discussion on Discord over #dev and #dev-public during the last week_

As e.g. the recent changes in middle game temp for T70 show, temperature has a significant effect on the draw rate, while the changes in eval are much smaller:

1. 0.9 until move 60, then 0.3 --> `e4` at start (702180) `D=0.214`, `WL=0.092`
2. 0.9 until move 40, decaying to 0.3 at move 60, then 0.3 --> `e4` at start (702392) `D=0.247`, `WL=0.093`
3. 0.9 until move 20, decaying to 0.3 at move 60. then 0.3 --> `e4` at start (702431) `D=0.317`, `WL=0.108` 

This means that the effects of temperature are nearly unbiased, while the effect on draw rate is much stronger. Closer inspection of the games show 2 reasons for that (as expected):

- small inaccuracies adding up, ultimately leading to a significant lead for one side, and
- one move blunders, which sometimes lead to an instant resignation.

Past attempts on ruling out one move blunders (bad game split, `--temp-visit-offset=-3`) were all detrimental, as the value head seems to need to see enough bad positions in training, and as recent temp experiments with early low endgame temp (move 26, decaying from 0 to 40) indicate, it's certainly not enough to only have them in the opening. However, we (arguably) want the small inaccuracies in all phases of the game as a signal of advantage and activity.

The following is the sketch of an idea which would limit the effects of one move blunders on evaluation (including `D` and `M`), but still allow the small inaccuracies, and is therefore different from effective temp 0 (like actual temp 0 or bad game split).

```
Whenever a move is picked to play, check for these 2 features:

1. was it the move with the highest N?
2. if not, was the difference in q to best_q below some threshold x?

Whenever either of these features is true, the move is accepted as a normal move.

When the selected move fails both criteria, it is considered a blunder,
and the game is effectively adjudicated and started new from the position after the move by

a) using `wl`, `d` and `m` of the best move as the game result for all positions until then
b) each blunder ends and starts such a "phase" of the game where 
   the "final" result is the adjudicated one
c) only the phase after the last blunder gets the actual game result
```

This is nearly doable with the current format of training data, and 100% compatible with what we use in actual NN training. The differences are:

- current game data have booleans for `win`. `draw`. `loss` instead of packed floats, while the format how `Q` of a position is stored would be needed
- while we save `best_q`, we don't save `chosen_q` (nor the number of visits for the chosen move or even the chosen move itself); however, this would be needed for triggering the above criteria, and it is not enough to compare it to the next position, because we then would treat "unknown" blunders (i.e. only turning out to be good after longer search) the same as "known" blunders where the move was known to be bad.
- we would also have to save the adjudicated values of `wl`, `q` and `d` for each phase and overwrite the current "real game endings"

From observing T70's evaluations for comparable training situations, a threshold of `x=0.2` or `x=0.3` seems like a rather good separation, and would very likely allow significantly higher endgame temperature without the current downsides.

**Comment by Naphthalin** [comment]
Note that this could be tested completely in SL, if game data was rescored appropriately whenever Q at root drops significantly from one move to another, or without data rescoring and using this directly as a not perfect approximation of "blundering knowingly".

**Comment by TesseractA** [comment]
What are the effects of a false signal, or a weak signal prior to the detection of the blunder? ``When the selected move fails both criteria, it is considered a blunder and the game is effectively adjudicated``

**Comment by Naphthalin** [comment]
What do you mean by "effects of a false/weak signal"?

**Comment by AlexisOlson** [comment]
I'm assuming @TesseractA meant what are the effects of adjudicating at points that were considered a "blunder" that are not actually blunders? It's possible that a "blunder" is actually a great move and early adjudication wouldn't let it learn that.

I'd guess that this wouldn't be a big concern in practice but I can't formally justify it one way or another.

**Comment by Naphthalin** [comment]
If I understand your question right, you mean a move which looks bad but in fact doesn't change the outcome of the game (not actually a blunder) or which actually even improves the position considerably. In both cases, the expectation is that by playing the move in training, the net learns to evaluate it better, and in the future the move will be evaluated better. In the first case, this means that it no longer triggers the q value threshold, and in the second case it actually is the best and most visited move. So in total, temp moves which trigger that "adjudication" wrongly are supposed to get their eval corrected and therefore not trigger the adjudication anymore.

**Comment by jhorthos** [comment]
What is best_q? Is that the highest q found for any of the moves (with the usual tree averaging)?

**Comment by jhorthos** [comment]
A perhaps related idea (suggested by aart) is to have the number of nodes explored for a move increase when Q for current move has changed more than some threshold from Q of the previous move.

**Comment by Naphthalin** [comment]
@jhorthos for the intentions of this issue, `best_q` would be q of best child, which would have been picked if it wasn't for temp.
The last idea sounds rather hacky, and at least to me it isn't clear whether this already should be reflected by KLD or not -- also, we shouldn't really need that in training, as a move having an unexpectedly high Q will improve the policy of that move.

**Comment by jhorthos** [comment]
Thanks, got it now. The Q change would be either direction, not just better, but anyway a different topic.

**Comment by RedDenver** [comment]
@Naphthalin Instead of adjudicating the game and effectively starting a new game, have you considered only training based on Q instead of Z (q_ratio=1, I think) for all moves before the blunder?

**Comment by Naphthalin** [comment]
> 
> 
> @Naphthalin Instead of adjudicating the game and effectively starting a new game, have you considered only training based on Q instead of Z (q_ratio=1, I think) for all moves before the blunder?

@RedDenver I have indeed considered that, and it was in fact my initial proposal. However, I felt that this would again introduce a too strong dependency on the settings, as with the preferred settings it's virtually impossible for the starting position to receive any training signal. Also, it basically solves the special case of "had a forced mate, but instead blundered via temp" and gives more accurate MLH values.

**Comment by Naphthalin** [comment]
After looking through the code, I believe the following parts of the code need to be changed:

- selfplay game data needs to log `WL`, `D` and `M` of the best move which would have been chosen without temperature, plus at least the `WL` value of the instead chosen move.
- training data format needs to be changed to contain the 3 best move's values as well, and the `game result` field needs to be changed from "ternary boolean" to a float 3-tupel of `{W,D,L}` identical to how `Q` is currently saved
- the converter from training data to NN training data needs a trivial change, as it doesn't need to translate the ternary boolean game result into a float 3-tupel anymore
- the rescorer needs an additional loop which takes a `blunder threshold` as input and works the following way: starting from the final move of the game, save the game result to a float 3-tupel `z={W,D,L}` and `m=M_adj`, and work backwards through the game. Whenever `best_WL - chosen_WL > blunder_threshold`, replace `z` and `m` with the `{W,D,L}` tupel resp. `M` of the best move, and set the `game result` of the position to the current `z` and `m` values. This could either be an additional script, or part of the TB rescorer, though I'd prefer an additional script for the below reason.

This way, we don't need to set the `blunder threshold` at training time, and it would be possible to download the training data and run the rescorer locally with variable blunder thresholds to create nets with different "optimism" about blunders in the actual game.

Additionally, it would in theory be possible to convert `blunder_threshold` and `temperature` into input parameters to the neural network similar to how KataGo is using the playout ratio as a NN input and allow a single neural network to cover a variety of "blunder optimism". However, this isn't needed for the above approach, and can be implemented later if the general strategy seems worth to expand in that direction.

**Comment by Naphthalin** [comment]
Road map to fully implement the described deblunder mechanism:

- introduce v6 training data format to include `played_q` (see #1468)
- include the rescorer logic into `loop.cc` of the `rescore_tb` branch of @Tilps (see my fork at https://github.com/Naphthalin/lc0/tree/rescore_tb )
- update `writer.cc`, possibly merge the added training data upgrading mechanism into `master`
- update several scripts on `lc0-training`, namely https://github.com/LeelaChessZero/lczero-training/blob/master/tf/chunkparser.py to populate the `winner` training data from `result_q`, `result_d` instead of `result` (called `winner` there)
- update v5 to v6 on all places I missed here

**Comment by Naphthalin** [comment]
Just mentioning it for completeness: The needed implementations were done, rescorer is updated on Tilps repo, and the T74 run testing it started in early February.

An idea still open is to make the deblunder trigger action probalistic (e.g. 0% for dQ < 0.05, 100% for dQ > 0.10, interpolate in between) to avoid an unintended potential thresholding possibly creating another bias.

**Comment by Naphthalin** [comment]
The interpolation also was done (not randomly, but with mixing Q and Z in the transition range), so closing this.
